README.md
# easy_dialogs

Easy to use dialogs with lists

- SingleChoiceDialog
- SingleChoiceConfirmationDialog

SingleChoiceConfirmationDialog 1           |  SingleChoiceConfirmationDialog 2 |  SingleChoiceDialog
:-------------------------:|:-------------------------:|:-------------------------:
![](art/single_choice_confirm_01.png)  |  ![](art/single_choice_confirm_02.png) | ![](art/single_choice_confirm_03.png)

SingleChoiceConfirmationDialog 1

```dart
SingleChoiceConfirmationDialog<String>(
            title: Text('Phone ringtone'),
            initialValue: _ringTone,
            items: _ringTones,
            onSelected: _onSelected,
            onSubmitted: _onSubmitted)
```

SingleChoiceConfirmationDialog 2

```dart
SingleChoiceConfirmationDialog<Color>(
            title: Text('Color'),
            initialValue: _color,
            items: _colors,
            contentPadding: EdgeInsets.symmetric(vertical: 16.0),
            divider: Container(
              height: 1.0,
              color: Colors.blue,
            ),
            onSubmitted: (Color color) {
              setState(() => _color = color);
            },
            itemBuilder: (Color color) =>
                Container(height: 100.0, color: color))
```

SingleChoiceDialog

```dart
SingleChoiceDialog<Color>(
              isDividerEnabled: true,
              title: Text('Pick a color'),
              items: _colors,
              onSelected: (Color color) {
                setState(() => _color = color);
              },
              itemBuilder: (Color color) =>
                  Container(height: 100.0, color: color))
```

CHANGELOG.md
## [2.0.1] - 9/20/2023
* Update to fix breaking changes

## [2.0.0] - 3/13/2021
* Add null-safety

## [1.1.0] - 3/7/2021
Deprecated stuff is replaced.
Thanks @dvorapa

## [1.0.0] - 9/9/2018

* Initial Release :tada:
* It includes [SingleChoiceDialog] and [SingleChoiceConfirmationDialog]


lib/single_choice_dialog.dart
import 'package:easy_dialogs/utils/dialog_utils.dart';
import 'package:easy_dialogs/utils/my_alert_dialog.dart';
import 'package:flutter/material.dart';

class SingleChoiceDialog<T> extends StatelessWidget {
  /// The choice items of type [T] to be used while building items of ListView
  /// which is content of the dialog
  final List<T> items;

  /// Callback that is called with selected item of type T when item is selected
  final ValueChanged<T>? onSelected;

  /// The (optional) title of the dialog is displayed in a large font at the top
  /// of the dialog.
  ///
  /// Typically a [Text] widget.
  final Widget? title;

  /// Padding around the title.
  ///
  /// If there is no title, no padding will be provided. Otherwise, this padding
  /// is used.
  ///
  /// This property defaults to providing 12 pixels on the top,
  /// 16 pixels on bottom of the title. If the [content] is not null, then no bottom padding is
  /// provided (but see [contentPadding]). If it _is_ null, then an extra 20
  /// pixels of bottom padding is added to separate the [title] from the
  /// [actions].
  final EdgeInsetsGeometry? titlePadding;

  /// Padding around the content.

  final EdgeInsetsGeometry contentPadding;

  /// The semantic label of the dialog used by accessibility frameworks to
  /// announce screen transitions when the dialog is opened and closed.
  ///
  /// If this label is not provided, a semantic label will be infered from the
  /// [title] if it is not null.  If there is no title, the label will be taken
  /// from [MaterialLocalizations.alertDialogLabel].
  ///
  /// See also:
  ///
  ///  * [SemanticsConfiguration.isRouteName], for a description of how this
  ///    value is used.
  final String? semanticLabel;

  ///Callback that is called with selected item of type T which returns a
  ///Widget to build list view item inside dialog
  final ItemBuilder<T>? itemBuilder;

  /// The (optional) horizontal separator used between title, content and
  /// actions.
  ///
  /// If this divider is not provided a [Divider] is used with [height]
  /// property is set to 0.0
  final Widget divider;

  /// The [divider] is not displayed if set to false. Default is set to true.
  final bool isDividerEnabled;

  SingleChoiceDialog(
      {Key? key,
      required this.items,
      this.onSelected,
      this.title,
      this.titlePadding,
      this.contentPadding = const EdgeInsets.fromLTRB(0.0, 12.0, 0.0, 16.0),
      this.semanticLabel,
      this.itemBuilder,
      this.isDividerEnabled = false,
      this.divider = const Divider(
        height: 0.0,
      )})
      : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MyAlertDialog(
      title: title,
      titlePadding: titlePadding,
      contentPadding: contentPadding,
      semanticLabel: semanticLabel,
      content: _buildContent(context),
      isDividerEnabled: isDividerEnabled,
      divider: divider,
    );
  }

  _buildContent(BuildContext context) {
    return ListView(
      shrinkWrap: true,
      children: items
          .map((item) => SimpleDialogOption(
                child: itemBuilder != null
                    ? itemBuilder!(item)
                    : Text(item.toString()),
                onPressed: () {
                  onSelected?.call(item);
                  Navigator.pop(context);
                },
              ))
          .toList(),
    );
  }
}


lib/utils/my_alert_dialog.dart
import 'package:flutter/material.dart';
import 'package:flutter/foundation.dart';

class MyAlertDialog<T> extends StatelessWidget {
  /// Creates an alert dialog.
  ///
  /// Typically used in conjunction with [showDialog].
  ///
  /// The [contentPadding] must not be null. The [titlePadding] defaults to
  /// null, which implies a default that depends on the values of the other
  /// properties. See the documentation of [titlePadding] for details.
  const MyAlertDialog({
    Key? key,
    this.title,
    this.titlePadding,
    this.content,
    this.contentPadding = const EdgeInsets.fromLTRB(24.0, 20.0, 24.0, 24.0),
    this.actions,
    this.semanticLabel,
    this.divider = const Divider(
      height: 0.0,
    ),
    this.isDividerEnabled = true,
  }) : super(key: key);

  /// The (optional) title of the dialog is displayed in a large font at the top
  /// of the dialog.
  ///
  /// Typically a [Text] widget.
  final Widget? title;

  /// Padding around the title.
  ///
  /// If there is no title, no padding will be provided. Otherwise, this padding
  /// is used.
  ///
  /// This property defaults to providing 24 pixels on the top, left, and right
  /// of the title. If the [content] is not null, then no bottom padding is
  /// provided (but see [contentPadding]). If it _is_ null, then an extra 20
  /// pixels of bottom padding is added to separate the [title] from the
  /// [actions].
  final EdgeInsetsGeometry? titlePadding;

  /// The (optional) content of the dialog is displayed in the center of the
  /// dialog in a lighter font.
  ///
  /// Typically, this is a [ListView] containing the contents of the dialog.
  /// Using a [ListView] ensures that the contents can scroll if they are too
  /// big to fit on the display.
  final Widget? content;

  /// Padding around the content.
  ///
  /// If there is no content, no padding will be provided. Otherwise, padding of
  /// 20 pixels is provided above the content to separate the content from the
  /// title, and padding of 24 pixels is provided on the left, right, and bottom
  /// to separate the content from the other edges of the dialog.
  final EdgeInsetsGeometry contentPadding;

  /// The (optional) set of actions that are displayed at the bottom of the
  /// dialog.
  ///
  /// Typically this is a list of [TextButton] widgets.
  ///
  /// These widgets will be wrapped in a [ButtonBar], which introduces 8 pixels
  /// of padding on each side.
  ///
  /// If the [title] is not null but the [content] _is_ null, then an extra 20
  /// pixels of padding is added above the [ButtonBar] to separate the [title]
  /// from the [actions].
  final List<Widget>? actions;

  /// The semantic label of the dialog used by accessibility frameworks to
  /// announce screen transitions when the dialog is opened and closed.
  ///
  /// If this label is not provided, a semantic label will be infered from the
  /// [title] if it is not null.  If there is no title, the label will be taken
  /// from [MaterialLocalizations.alertDialogLabel].
  ///
  /// See also:
  ///
  ///  * [SemanticsConfiguration.isRouteName], for a description of how this
  ///    value is used.
  final String? semanticLabel;

  /// The (optional) horizontal separator used between title, content and
  /// actions.
  ///
  /// If this divider is not provided a [Divider] is used with [height]
  /// property is set to 0.0
  final Widget divider;

  ///[divider] is not displayed if set to false. Default is set to true.
  final bool isDividerEnabled;

  @override
  Widget build(BuildContext context) {
    final List<Widget> children = <Widget>[];
    String? label = semanticLabel;

    if (title != null) {
      children.add(new Padding(
        padding: titlePadding ??
            new EdgeInsets.fromLTRB(
                24.0, 24.0, 24.0, isDividerEnabled ? 20.0 : 0.0),
        child: new DefaultTextStyle(
          style: Theme.of(context).textTheme.titleLarge!,
          child: new Semantics(child: title, namesRoute: true),
        ),
      ));
      if (isDividerEnabled) children.add(divider);
    } else {
      switch (defaultTargetPlatform) {
        case TargetPlatform.iOS:
        case TargetPlatform.macOS:
          break;
        case TargetPlatform.android:
        case TargetPlatform.fuchsia:
        case TargetPlatform.linux:
        case TargetPlatform.windows:
          label = semanticLabel ??
              MaterialLocalizations.of(context).alertDialogLabel;
      }
    }

    if (content != null) {
      children.add(new Flexible(
        child: Padding(
          padding: contentPadding,
          child: DefaultTextStyle(
            style: Theme.of(context).textTheme.titleMedium!,
            child: content!,
          ),
        ),
      ));
    }

    if (actions != null) {
      if (isDividerEnabled) children.add(divider);
      children.add(ButtonBarTheme(
        data: Theme.of(context).buttonBarTheme,
        child: ButtonBar(
          children: actions!,
        ),
      ));
    }

    Widget dialogChild = Column(
      mainAxisSize: MainAxisSize.min,
      crossAxisAlignment: CrossAxisAlignment.stretch,
      children: children,
    );

    if (label != null)
      dialogChild =
          Semantics(namesRoute: true, label: label, child: dialogChild);

    return Dialog(child: dialogChild);
  }
}


lib/single_choice_confirmation_dialog.dart
import 'package:easy_dialogs/utils/dialog_utils.dart';
import 'package:easy_dialogs/utils/my_alert_dialog.dart';
import 'package:flutter/material.dart';

class SingleChoiceConfirmationDialog<T> extends StatefulWidget {
  /// The initial selected item when Dialog is opened.

  /// The  groupValue of [RadioListTile]'s inside [ListView] is set to this
  /// value initially

  /// If not provided, no RadioListTile is selected.

  final T? initialValue;

  /// The choice items of type [T] to be used while building items of ListView
  /// which is content of the dialog
  final List<T> items;

  /// Callback that is called with selected item of type T when item is selected
  final ValueChanged<T>? onSelected;

  /// Callback that is called with selected item of type T when submit button
  /// inside actions is tapped.
  ///
  /// This is only called when default actions are used.
  ///
  /// If [actions] are provided, then user should save the value when
  /// [onSelected] is called and handle the submit with this value inside a
  /// submit button of this actions.
  final ValueChanged<T>? onSubmitted;

  /// The (optional) title of the dialog is displayed in a large font at the top
  /// of the dialog.
  ///
  /// Typically a [Text] widget.
  final Widget? title;

  /// Padding around the title.
  ///
  /// If there is no title, no padding will be provided. Otherwise, this padding
  /// is used.
  ///
  /// This property defaults to providing 12 pixels on the top,
  /// 16 pixels on bottom of the title. If the [content] is not null, then no bottom padding is
  /// provided (but see [contentPadding]). If it _is_ null, then an extra 20
  /// pixels of bottom padding is added to separate the [title] from the
  /// [actions].
  final EdgeInsetsGeometry? titlePadding;

  /// Padding around the content.

  final EdgeInsetsGeometry contentPadding;

  /// The semantic label of the dialog used by accessibility frameworks to
  /// announce screen transitions when the dialog is opened and closed.
  ///
  /// If this label is not provided, a semantic label will be infered from the
  /// [title] if it is not null.  If there is no title, the label will be taken
  /// from [MaterialLocalizations.alertDialogLabel].
  ///
  /// See also:
  ///
  ///  * [SemanticsConfiguration.isRouteName], for a description of how this
  ///    value is used.
  final String? semanticLabel;

  /// Callback that is called with selected item of type T which returns a
  /// Widget to build list view item inside dialog
  final ItemBuilder<T>? itemBuilder;

  /// The (optional) set of actions that are displayed at the bottom of the
  /// dialog.
  ///
  /// Typically this is a list of [TextButton] widgets.
  ///
  /// These widgets will be wrapped in a [ButtonBar], which introduces 8 pixels
  /// of padding on each side.
  ///
  /// If the [title] is not null but the [content] _is_ null, then an extra 20
  /// pixels of padding is added above the [ButtonBar] to separate the [title]
  /// from the [actions].
  final List<Widget>? actions;

  /// The color of the selected item's radio
  final Color? activeColor;

  /// The label of cancel button inside actions
  final String? cancelActionButtonLabel;

  /// The label of submit button inside actions
  final String? submitActionButtonLabel;

  /// The color of label of the default action buttons
  final Color? actionButtonLabelColor;

  /// The (optional) horizontal separator used between title, content and
  /// actions.
  ///
  /// If this divider is not provided a [Divider] is used with [height]
  /// property is set to 0.0
  final Widget divider;

  SingleChoiceConfirmationDialog({
    Key? key,
    this.initialValue,
    required this.items,
    this.onSelected,
    this.onSubmitted,
    this.title,
    this.titlePadding,
    this.contentPadding = const EdgeInsets.fromLTRB(0.0, 0.0, 0.0, 0.0),
    this.semanticLabel,
    this.actions,
    this.itemBuilder,
    this.activeColor,
    this.cancelActionButtonLabel,
    this.submitActionButtonLabel,
    this.actionButtonLabelColor,
    this.divider = const Divider(height: 0.0),
  }) : super(key: key);

  @override
  _SingleChoiceConfirmationDialogState<T> createState() =>
      _SingleChoiceConfirmationDialogState<T>();
}

class _SingleChoiceConfirmationDialogState<T>
    extends State<SingleChoiceConfirmationDialog<T>> {
  T? _chosenItem;

  @override
  void initState() {
    _chosenItem = widget.initialValue;
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return MyAlertDialog(
      title: widget.title,
      titlePadding: widget.titlePadding,
      contentPadding: widget.contentPadding,
      semanticLabel: widget.semanticLabel,
      content: _buildContent(),
      actions: _buildActions(),
      divider: widget.divider,
    );
  }

  _buildContent() {
    return ListView(
      shrinkWrap: true,
      children: widget.items
          .map(
            (item) => RadioListTile<T>(
              title: widget.itemBuilder != null
                  ? widget.itemBuilder!(item)
                  : Text(item.toString()),
              activeColor: widget.activeColor ?? Theme.of(context).colorScheme.secondary,
              value: item,
              groupValue: _chosenItem,
              onChanged: (value) {
                if (value != null) {
                  widget.onSelected?.call(value);
                  setState(() {
                    _chosenItem = value;
                  });
                }
              },
            ),
          )
          .toList(),
    );
  }

  _buildActions() {
    return widget.actions ??
        <Widget>[
          TextButton(
            style: TextButton.styleFrom(
                foregroundColor: widget.actionButtonLabelColor ??
                    Theme.of(context).colorScheme.secondary),
            child: Text(widget.cancelActionButtonLabel ?? 'CANCEL'),
            onPressed: () {
              Navigator.pop(context);
            },
          ),
          TextButton(
            style: TextButton.styleFrom(
                foregroundColor: widget.actionButtonLabelColor ??
                    Theme.of(context).colorScheme.secondary),
            child: Text(widget.submitActionButtonLabel ?? 'OK'),
            onPressed: () {
              Navigator.pop(context);
              if (_chosenItem != null) {
                widget.onSubmitted?.call(_chosenItem!);
              }
            },
          )
        ];
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:easy_dialogs/easy_dialogs.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Easy Dialogs Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      routes: {
        '/': (context) => HomePage(),
      },
    );
  }
}

class HomePage extends StatefulWidget {
  @override
  HomePageState createState() {
    return new HomePageState();
  }
}

class HomePageState extends State<HomePage> {
  final List<String> _ringTones = [
    'None',
    'Callisto',
    'Ganymede',
    'Luna',
    'Oberon',
    'Phobos',
    'Dione',
    'Jungle Gym',
    'Ukulele',
    'Snowflakes',
  ];

  final List<Color> _colors = [
    Colors.orange,
    Colors.pinkAccent,
    Colors.black,
    Colors.blue,
    Colors.brown,
    Colors.green,
  ];

  String _ringTone = "None";

  Color _color = Colors.black;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Settings'),
      ),
      body: ListView(
        children: <Widget>[
          ListTile(
            onTap: _openRingtoneDialog,
            title: Text('Ringtone'),
            subtitle: Text(_ringTone),
            leading: Icon(Icons.music_note),
          ),
          ListTile(
              onTap: _openColorDialog,
              title: Text('Color'),
              subtitle: Container(
                margin: EdgeInsets.only(top: 8.0),
                height: 20.0,
                color: _color,
              ),
              leading: Icon(Icons.colorize)),
          ListTile(
              onTap: _openColorDialogWithoutConfirmation,
              title: Text('Color without confirmation'),
              subtitle: Container(
                margin: EdgeInsets.only(top: 8.0),
                height: 20.0,
                color: _color,
              ),
              leading: Icon(Icons.colorize)),
        ],
      ),
    );
  }

  _openRingtoneDialog() {
    showDialog(
        context: context,
        builder: (context) => SingleChoiceConfirmationDialog<String>(
            title: Text('Phone ringtone'),
            initialValue: _ringTone,
            items: _ringTones,
            onSelected: _onSelected,
            onSubmitted: _onSubmitted));
  }

  _openColorDialog() {
    showDialog(
      context: context,
      builder: (context) => SingleChoiceConfirmationDialog<Color>(
            title: Text('Color'),
            initialValue: _color,
            items: _colors,
            contentPadding: EdgeInsets.symmetric(vertical: 16.0),
            divider: Container(
              height: 1.0,
              color: Colors.blue,
            ),
            onSubmitted: (Color color) {
              setState(() => _color = color);
            },
            itemBuilder: (Color color) =>
                Container(height: 100.0, color: color),
          ),
    );
  }

  _openColorDialogWithoutConfirmation() {
    showDialog(
        context: context,
        builder: (context) => SingleChoiceDialog<Color>(
              isDividerEnabled: true,
              title: Text('Pick a color'),
              items: _colors,
              onSelected: (Color color) {
                setState(() => _color = color);
              },
              itemBuilder: (Color color) =>
                  Container(height: 100.0, color: color),
            ));
  }

  void _onSelected(String value) {
    print('Selected $value');
    setState(() {
      _ringTone = value;
    });
  }

  void _onSubmitted(String value) {
    print('Submitted $value');
    setState(() {
      _ringTone = value;
    });
  }
}
